File: algorithms/rs/239.滑动窗口最大值.rs
/*
 * @lc app=leetcode.cn id=239 lang=rust
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
use std::collections::VecDeque;
impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let len = nums.len();
        if k == 1 {
            return nums;
        }
        let mut de_queue = VecDeque::with_capacity(k as usize);
        let mut ans = vec![];
        for i in (0 as usize)..len {
            Solution::push(&mut de_queue, nums[i]);
            if i as i32 == k - 1 {
                ans.push(Solution::max(&de_queue))
            } else if i as i32 > k - 1 {
                Solution::pop(&mut de_queue, nums[i - k as usize]);
                ans.push(Solution::max(&de_queue))
            }
        }
        ans
    }
    fn max(dequeue: &VecDeque<i32>) -> i32 {
        *dequeue.front().unwrap()
    }
    fn push(dequeue: &mut VecDeque<i32>, v: i32) {
        while !dequeue.is_empty() && *dequeue.back().unwrap() < v {
            dequeue.pop_back();
        }
        dequeue.push_back(v);
    }
    fn pop(dequeue: &mut VecDeque<i32>, v: i32) {
        if *dequeue.front().unwrap() == v {
            dequeue.pop_front();
        }
    }
}
// @lc code=end

File: algorithms/golang/239.滑动窗口最大值.go
/*
 * @lc app=leetcode.cn id=239 lang=golang
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
/**
 * 单调递减栈
 * 每次窗口移动，弹出窗口外元素，新元素从栈顶开始与栈内元素碰撞，小于新元素的值全部弹出，最后新元素入栈。
 * 单调栈底部元素则为当前滑动窗口最大元素。
 * 时间复杂度O(n)，空间复杂度O(k)。
 */
func maxSlidingWindow(nums []int, k int) []int {
	ans := []int{}
	queue := make([]int, 0)
	for i, val := range nums {
		if i >= k && len(queue) > 0 && queue[0] == nums[i-k] {
			queue = queue[1:]
		}
		for len(queue) > 0 && queue[len(queue)-1] < val {
			queue = queue[:len(queue)-1]
		}
		queue = append(queue, val)
		if i >= k-1 {
			ans = append(ans, queue[0])
		}
	}
	return ans
}

// @lc code=end

